packages/mui-base/src/useNumberField/useNumberField.ts

```diff
@@ -0,0 +1,802 @@
+
+        // If the input value is invalid, set it to the last valid value.
+        if (parsedValue !== null) {
+          setValue(parsedValue);
+        }
+      },
+      onChange(event) {
+        externalProps.onChange?.(event);
+        // Workaround for https://github.com/facebook/react/issues/9023
+        if (event.nativeEvent.defaultPrevented) {
+          return;
+        }
+
+        allowInputSyncRef.current = false;
+
+        const targetValue = event.target.value;
+        setInputValue(event.target.value);
+
+        if (event.isTrusted) {
+          return;
+        }
+
+        if (targetValue.trim() === '') {
+          setValue(null);
+          return;
+        }
+
+        const parsedValue = parseNumber(targetValue, formatOptionsRef.current);
+
+        if (parsedValue !== null) {
+          setValue(parsedValue);
+        }
+      },
+      onKeyDown(event) {
+        externalProps.onKeyDown?.(event);
+        if (event.defaultPrevented || readOnly) {
+          return;
+        }
+
+        allowInputSyncRef.current = true;
+
+        let isAllowedNonNumericKey = allowedNonNumericKeys.includes(event.key);
+
+        const { decimal } = getNumberLocaleDetails(formatOptionsRef.current);
+
+        // Allow the minus key only if there isn't already a plus or minus sign.
+        if (event.key === '-') {
+          isAllowedNonNumericKey = !inputValue.includes('-');
+        }
+
+        // Allow only one decimal separator.
+        if (event.key === decimal) {
+          isAllowedNonNumericKey = !inputValue.includes(decimal);
+        }
+
+        const arabicNumeralsRegex = new RegExp(`^[${ARABIC_NUMERALS.join('')}]$`);
+        const hanNumeralsRegex = new RegExp(`^[${HAN_NUMERALS.join('')}]$`);
+        const isLatinNumeral = /^[0-9]$/.test(event.key);
+        const isArabicNumeral = arabicNumeralsRegex.test(event.key);
+        const isHanNumeral = hanNumeralsRegex.test(event.key);
+        const isNavigateKey = [
+          'Backspace',
+          'Delete',
+          'ArrowLeft',
+          'ArrowRight',
+          'Home',
+          'End',
+          'Tab',
+          'Enter',
+        ].includes(event.key);
+
+        if (
+          // Allow composition events (e.g., pinyin)
+          event.nativeEvent.isComposing ||
+          event.altKey ||
+          event.ctrlKey ||
+          event.metaKey ||
+          isAllowedNonNumericKey ||
+          isLatinNumeral ||
+          isArabicNumeral ||
+          isHanNumeral ||
+          isNavigateKey
+        ) {
+          return;
+        }
+
+        // We need to commit the number at this point if the input hasn't been blurred.
+        const parsedValue = parseNumber(inputValue, formatOptionsRef.current);
+        if (parsedValue !== null) {
+          rawValueRef.current = parsedValue;
+        }
+
+        event.preventDefault();
+
+        const amount = getStepAmount();
+
+        if (event.key === 'ArrowUp') {
+          event.stopPropagation();
```

Comment:
Should we do this? It tends to be a foot gun https://css-tricks.com/dangers-stopping-event-propagation/

```suggestion
```

We rarely call this in the codebase, and always add a comment for why when we do. It makes me think of https://twitter.com/jjenzz/status/1767736778205630646

Reply:
For arrow keys specifically, I think yes. If you have bound arrow key listeners on the window to execute some functionality (e.g. shortcuts for actions), you don't want to propagate the event to them when using the number field.

If they want to always execute those, they can use `capture` event listeners to workaround it. I stop propagation for arrow events in Floating UI, and there has yet to be an issue about it.

Reply:
Actually, nevermind on this. The `input` element automatically stops propagation itself, so there's no need to call it on `keydown`.